### ISSUE TITLE
unused_parens false positive on cfg_attr(..., derive(..., ...))

### ISSUE THREAD

**Issue by behnam** [rust-lang/rust#48871]
UNIC builds started failing today on nightly with these errors:

```rust
error: unnecessary parentheses around function argument
  --> unic/bidi/src/level.rs:34:49
   |
34 | #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
   |                                                 ^^^^^^^^^^^
   |
note: lint level defined here
  --> unic/bidi/src/lib.rs:14:20
   |
14 | #![deny(bad_style, unused)]
   |                    ^^^^^^
   = note: #[deny(unused_parens)] implied by #[deny(unused)]
help: remove these parentheses
   |
34 | #[cfg_attr(feature = "serde", derive(Serialize, match <u8 as _serde::Deserialize>::deserialize(__e) {
35 |      ::result::Result::Ok(val) => val,
36 |      ::result::Result::Err(err) => {
37 |          return ::result::Result::Err(::convert::From::from(err))
38 |      }
39 |  }))]
   |
error: unnecessary parentheses around function argument
  --> unic/bidi/src/level.rs:34:49
   |
34 | #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
   |                                                 ^^^^^^^^^^^ help: remove these parentheses
error: aborting due to 2 previous errors
```
(from <https://travis-ci.org/behnam/rust-unic/jobs/351099208>)

Problems:
- the `cfg_attr` lines do not have any unused parentheses, so it's a false positive;
- the location the error message is pointing to is misleading, as there's no parentheses in the span it's pointing to; and
- there are two different error reports here, both pointing to the same symptom, but with different looks.

**Comment by zackmdavis** [comment]
Possibly-relevant context: the unnecessary-parens lint was extended to arguments in #46980, and neutered for arguments in nested macros in #47896. But if this started failing _today_, something relevant must have changed elsewhere to reveal this bug.

**Comment by ExpHP** [comment]
Or it could be a change in `serde_derive`. [This recent commit](https://github.com/serde-rs/serde/commit/30361ac6d022839be19fc608645756ed1be839ab) looks interesting, but it actually *removes* parentheses instead of adding them.  What happens if you try `serde` on git master?

BTW, I tried building unic locally, and was unable to reproduce this with unic@`9466e05e`, rustc `nightly-2018-03-07`, and any of `serde_derive` versions `1.0.28`-`1.0.30` (the latest).

**Comment by nikomatsakis** [comment]
Hello from the @rust-lang/compiler triage meeting.

@benham can you verify if you are able to reproduce? If so, maybe give a specific commit and the output from `rustc -vV`? It'd be great if we can isolate down what has changed.

**Comment by nikomatsakis** [comment]
triage: P-high

(But no assignment until we can reproduce.)

**Comment by behnam** [comment]
Thanks, @nikomatsakis, for the follow up!

Looks at the [UNIC build history](https://travis-ci.org/behnam/rust-unic/builds), apparently it was only failing for one day, on nightly, and got fixed. The version for that nightly build was `rustc 1.26.0-nightly (2789b067d 2018-03-06)`.

I reported the issue right away to help with triage. Since it hasn't been repro'ed for the past 9 days, please feel free to close it.

**Comment by nikomatsakis** [comment]
Thanks!
